docs/rfcs/20201208_fragmentation.md

```diff
@@ -0,0 +1,303 @@
+
+Once a large allocation is detected it, avoiding it depends on the context (browsing the related ScyllaDB issues shows the large variety of approaches taken). Sometimes it is as simple as using a fragmented data structure (linked list, iobuf, deque) rather than a contiguous one. In other cases it was not expected that such a structure became so large in the first place, so some other adjustment is needed.
+
+#### Drawbacks
+
+A significant downside of the this approach is that is mostly reactive: problems are identified only when they cause a crash or (hopefully) when the warning is detected and acted on. Although the issue may be resolved, significant downtime may already have been incurred.
+
+Another issue is that many allocations occur in third party libraries rather than the core code: although we have the source to all third parties and can patch them, this creates a divergence with upstream, creates a maintenance burden, and so on. Furthermore, it may be harder to avoid large allocations in third party libraries since we won't the same tools available to use as 
+
+#### Recommendations
+
+At least, we should enable the large allocation warning, perhaps at 256 KiB. We can have it enabled in testing with a post-processing phase which determines if the discovered large allocations are from new locations which need to vetted to check how large they can get.
+
+We should expose metrics related to fragmentation such as availability % for certain sizes in our metrics and possibly in telemetry, in order to understand what "typical" fragmentation looks like.
+
+### Make Reclaim Span-Aware
+
+As described above, the current reclaim mechanism means that reclaimers aren't aware of the required *size* for the pending allocation. So they may repeatedly free scattered chunks of memory which can't satisfy the current allocation. In the worst case, we can free almost the entire batch cache until we finally get a large enough contiguous chunk of memory. This means that fragmentation can introduce a tricky "semi-failure" mode rather than total failure (failed allocations). In this mode, performance is badly degraded because the batch cache is continually forced to a small size in order to satisfy large allocations inefficiently.
+
+To mitigate this, we could implement a span-aware reclaimer: the reclaim code tells the reclaimer the size of the allocation and the reclaimer tries to free spans that will result in a contiguous region of at least the given size. For example, rather than LRU eviction, use an eviction algorithm which evicts entries which happen to be contiguous, or perhaps a hybrid algorithm which considers both the contiguity and age of the entries. This could significantly reduce the impact of fragmentation since the batch cache would suffer fewer evictions while reclaiming.
```

Comment:
> tells the reclaimer the size of the allocation and the reclaimer tries to free spans that will result in a contiguous region of at least the given size.

IIRC the reclaimer does receive a target size as part of the up call indicating how much space is needed. Am I remembering that incorrectly or maybe you are proposing a different type of metadata?

Reply:
Yeah, that part perhaps wasn't clear enough: the allocator does already tell the reclaimer how much to free: the suggestion change here is that the reclaimer should try to free that much _contiguous_ space, not just that much space total. Currently, the allocator says "hey, I want 8 MB _contiguous_" and the relcaimer then frees 8 MB scattered around randomly, not really satisfying the request. So this change would be in the reclaimer, mostly (though how to support this efficiently is an open question since it means the reclaim operation is basically an optimization problem now).